include/onnxruntime/core/providers/cuda/cuda_provider_factory.h

```diff
@@ -11,6 +11,7 @@ extern "C" {
  * \param device_id cuda device id, starts from zero.
  */
 ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_CUDA, _In_ OrtSessionOptions* options, int device_id);
+ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_CUDA_CONV_ALGO, _In_ OrtSessionOptions* options, int device_id, int conv_algo);
```

Comment:
It'll be better to have an extensible way to configure the EP. This will avoid the need to add a new API for every new configuration. OpenVINO and Nuphar use a serialized settings string. Examples: https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/providers/openvino/openvino_provider_factory.h#L23 and https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/providers/nuphar/nuphar_provider_factory.h#L13.

Reply:
I feel it's fine. We don't provide binary compatibility for the other EPs. Because they have dependencies, their dependencies doesn't support binary compatibility, then it's meanless for us to try to provide such things. I would prefer to modify OrtSessionOptionsAppendExecutionProvider_CUDA directly.

Reply:
Interesting, I agree that OrtSessionOptionsAppendExecutionProvider_CUDA is not part of the ORT versioned API but it sounds like anyone using this function needs a custom build and it's non ABI?

Reply:
OrtSessionOptionsAppendExecutionProvider_* functions may not be part of the versioned struct, but is very much part of the C interface ORT exports and should be ABI. When did we change this? Not sure why we care about the ABI guarantees of the dependencies here; this is about what ORT exposes through its C APIs.
CUDA doesn't need a custom build. We provide a pre-built package for CUDA EP.

Reply:
@pranavsharma, the binary compatible promise should only apply to the CPU only build. Because there is no way of us to provide such things for CUDA EP enabled build.

Reply:
We provide an official cuda ep build and this function is exported by ORT DLL in our official binaries. Not following why then we shouldn't maintain ABI compatibility? Our users are using the ORT DLL, not the cuda dlls.

Reply:
Binary compatible  means if an application was built with ORT 1.4, then it can continue to be used with ORT 1.5 without rebuild, the user can just replace onnxruntime.dll with a new version.  I think it doesn't apply to CUDA cases. ORT 1.4 depends on CUDA 10.1, ORT 1.5 depends on CUDA 10.2, anyway it doesn't work directly. And no user ask for such a feature.

Reply:
If you install the right cuda deps with a newer ORT dll, users should still expect previous ORT exported APIs to work as before. I thought that was the whole point of ABIs. If we remove this API, people who will upgrade will start seeing failures. If we don't want to maintain this guarantee for any reason, it should be clearly documented in the header file. I don't see any such documentation.

Reply:
@pranavsharma You're right. Thank you!

Reply:
Done.